### PR TITLE
pr-2019-03-28

### PR DESCRIPTION
**Pull request by jpdunc23**
Merges in feature branch [jpdunc23/nCompiler/logging](https://github.com/jpdunc23/nCompiler/tree/logging)

### CODE REVIEW

**Comment by jpdunc23** [comment]
@perrydv see https://github.com/jpdunc23/nCompiler/commit/c0d6dafabf6fb2cf08831cecac2c81b25826e6c9

I had to compromise because I wasn't able to make the `nOption<-` idiom work... I believe the idiom can't support assigning in environments or perhaps I'm doing something wrong.

**Comment by jpdunc23** [comment]
Merge in feature branch [jpdunc23/nCompiler/math-testing](https://github.com/jpdunc23/nCompiler/tree/math-testing)

**Comment by jpdunc23** [comment]
Merge in feature branch [jpdunc23/nCompiler/if-else-while-for](https://github.com/jpdunc23/nCompiler/tree/if-else-while-for)

https://github.com/nimble-dev/nCompiler/pull/3/commits/393be9955097c6a6d1061b5aa0a19d1ec4d941b0 adds a minor bugfix to `AssignAfterRecursing` where LHS of `<-` was inheriting isArg = TRUE when RHS was one of the arguments. See https://github.com/jpdunc23/nCompiler/commit/47ee8270a313805d97925875f464c6c1ebc69cd7

File: nCompiler/R/NF_CompilerClass.R [hunk 1]
```diff
@@ -142,16 +142,36 @@ processNFstages <- function(NFcompiler,
     )
     debug <- controlFull$debug
     debugCpp <- FALSE##controlFull$debugCpp
+    loggingOpt <- get_nOption('compilerOptions')[['logging']]
+    logging <- controlFull$logging || loggingOpt
+    if (logging && !loggingOpt) {
+      ## to make logging option accessible in handlers, must set TRUE in .nOptions
+      set_nOption('logging', TRUE, 'compilerOptions')
```
Comment (perrydv) [inline nCompiler/R/NF_CompilerClass.R]: A common approach would be to include `on.exit(set_nOption('logging', FALSE, 'compilerOptions'))`.  That way, whether exit is by error or regular return, the option will always revert to its original setting.

File: nCompiler/R/compile_eigenization.R [hunk 2]
```diff
@@ -143,10 +143,10 @@ compile_eigenize <- function(code,
             return(setupExprs)
         }
         if(code$name == '[') { ## If there is still A[i] in the code, it is because it is equivalent to a scalar and does not need eigenization
-            if(code$nDim == 0) {
+            if(code$type$nDim == 0) {
                 return(NULL)
             } else {
-                writeLines(paste0('Warning, in eigenizing ',nimDeparse(code), ' the [ is still there but nDim is not 0 (not a scalar).') )
+                writeLines(paste0('Warning, in eigenizing ',deparse(code$Rexpr), ' the [ is still there but nDim is not 0 (not a scalar).') )
```
Comment (perrydv) [inline nCompiler/R/compile_eigenization.R]: I think this should be `nDeparse(code)`

File: nCompiler/R/compile_exprClass.R [hunk 3]
```diff
@@ -148,8 +148,8 @@ addIndentToList <- function(x, indent) {
 
 ## error trapping utilities to be used from the various processing steps
 exprClassProcessingErrorMsg <- function(code, msg) {
-    contextCode <- if(!is.null(code$caller)) paste(unlist(nimDeparse(code$caller)), collapse = '\n') else character()
-    ans <- paste0(msg, '\n This occurred for: ', nimDeparse(code),'\n', collapse = '')
+    contextCode <- if(!is.null(code$caller)) deparse(code$caller$Rexpr) else character()
```
Comment (perrydv) [inline nCompiler/R/compile_exprClass.R]: Same as above: I think we want to use `nDeparse(code)`.  Maybe not.  Let's discuss.

File: nCompiler/R/NF_Compile.R [hunk 4]
```diff
@@ -25,6 +25,8 @@ nCompile_nFunction <- function(NF,
         get_nOption('compilerOptions'),
         control
     )
+    loggingOpt <- get_nOption('compilerOptions')[['logging']]
+    logging <- controlFull$logging || loggingOpt
```
[the remark this answers is not in the document]
Comment (jpdunc23) [reply nCompiler/R/NF_Compile.R]: Agreed. Updated here: https://github.com/nimble-dev/nCompiler/pull/3/commits/418ae093565696c7652caadb97c2662128482183

File: nCompiler/R/NF_CompilerClass.R (hunk 1 above)
Comment (jpdunc23) [reply nCompiler/R/NF_CompilerClass.R]: Thanks! Updated here: 418ae09

File: nCompiler/R/compile_eigenization.R (hunk 2 above)
Comment (jpdunc23) [reply nCompiler/R/compile_eigenization.R]: Updated: https://github.com/nimble-dev/nCompiler/pull/3/commits/749aaf11b87d1a69a2947ce2c021a997451d061d

File: nCompiler/R/compile_exprClass.R (hunk 3 above)
Comment (jpdunc23) [reply nCompiler/R/compile_exprClass.R]: Updated, using nDeparse for now: https://github.com/nimble-dev/nCompiler/pull/3/commits/749aaf11b87d1a69a2947ce2c021a997451d061d

**Pull request merged**